chore: implementation skill (#1149)
k0rdent/kof: .agents/skills/implementation/SKILL.md

### 3. PromxyServerGroupReconciler
Watches `PromxyServerGroup` CRs. Aggregates them by `secret-name` label, renders promxy config, updates the promxy Secret, and POSTs `/-/reload`.

**Impact**: spec changes must be reflected in the config template under `kof-operator/internal/controller/`.

### 4. AlertsConfigMapReconciler
Watches alert/record rule ConfigMaps. Merges `PrometheusRule` CRs and per-cluster ConfigMaps into the mothership promxy rules Secret and per-cluster vmrules ConfigMaps (injected into `kof-storage` via MCS).

**Impact**: `defaultRules.*` in `charts/kof-mothership/values.yaml` controls which rule groups are enabled.

### 5. VMStorageConnectionReconciler
Watches `VMStorageConnection` CRs. Patches `VTCluster`/`VLCluster` extraArgs to add storage node addresses. Uses a finalizer for clean deletion.

**Impact**: mothership VLCluster/VTCluster must exist before regional clusters register — `kof-mothership` must deploy first.

---

## Helm Chart Dependency Chain

### The `charts/kof` Umbrella Chart

`charts/kof` is the single entry point — it creates FluxCD HelmRelease objects for all components. No workloads run in this chart itself.

> 

Reviewer: ```suggestion
Some charts like `charts/cold-storage-exporter` are not registered as `kof` components, so they may or may not be installed manually.

```

**Install order** is defined by `dependsOn` per component in `charts/kof/values.yaml`. When adding a new chart, add it to `global.components` and set its `dependsOn`.

### Chart-to-Cluster Deployment Map

| Chart | Deployed Where | How |
|---|---|---|
| `kof-operators` | Mothership | FluxCD HelmRelease |
| `victoria-metrics-operator` | Mothership | FluxCD HelmRelease |
| `kof-mothership` | Mothership | FluxCD HelmRelease |
| `kof-dashboards` | Mothership | Bundled dependency of kof-mothership and kof-storage |
| `kof-regional` | Mothership | HelmRelease — creates MCS policy objects only |
| `kof-child` | Mothership | HelmRelease — creates MCS policy objects only |
| `kof-storage` | Regional clusters | MCS rendered by kof-regional |
| `kof-collectors` | Regional and child clusters | MCS rendered by kof-regional and kof-child |
| `kof-propagation` | Mothership | ServiceTemplate for ConfigMap propagation |
| `audit-logs-exporter` | Regional clusters | Dependency of kof-storage |

> | `cold-storage-exporter` | Regional/mothership | Dependency of kof-storage |

Reviewer: ```suggestion
| `audit-logs-exporter` | Regional clusters | Optional manual helm install |
| `cold-storage-exporter` | Regional/mothership | Optional manual helm install |
```

### Two-Layer Helm Templating in MCS Values

> `kof-regional` and `kof-child` each render a `MultiClusterService`. Their `services[].values` use **Helm-within-Helm templates**: the outer layer runs at mothership install time (bakes in umbrella values and feature flags), the inner layer runs at Sveltos apply time (reads live cluster annotations and referenced Secrets).

Reviewer: ```suggestion
`kof-regional` and `kof-child` each render a `MultiClusterService`. Their `services[].values` use **Sveltos-within-Helm templates**: the outer layer runs at mothership install time (bakes in umbrella values and feature flags), the inner layer runs at Sveltos apply time (reads live cluster annotations and referenced Secrets).
```

```

Operators can override any collector or storage value per cluster via a JSON annotation on `ClusterDeployment` (e.g., `k0rdent.mirantis.com/kof-collectors-values`).

### Values Consistency

A CI script in `scripts/` validates that every default in `charts/kof/values.yaml` exists at the same path in the component's own chart. Run it after changing defaults in either location.

---

## KOF-Native CRDs (`kof-operator/api/v1beta1/`, group `kof.k0rdent.mirantis.com/v1beta1`)

**PromxyServerGroup** — describes one remote Prometheus-compatible backend for promxy. Key fields: `targets`, `pathPrefix`, `scheme`, `clusterName`, `basicAuth`, `tlsConfig`. The `k0rdent.mirantis.com/secret-name` label groups multiple objects into one promxy config Secret.

**VMStorageConnection** — registers a remote VictoriaMetrics storage node with a `VTCluster` or `VLCluster`. Key fields: `clusterRef`, `address`, `authSecret`, `tlsInsecureSkipVerify`.

---

## Multi-Cluster Propagation Pattern

1. Operator creates a `MultiClusterService` in `kcm-system`.
2. MCS carries a cluster selector and `services` list with `ServiceTemplate` references.

> 3. KCM/Sveltos renders and applies templates to matching clusters.

Reviewer: ```suggestion
3. KSM (part of KCM) / Sveltos renders and applies templates to matching clusters.
```

4. `templateResourceRefs` injects ConfigMap/Secret values from the mothership at render time.

**Naming conventions**:
- Regional ConfigMap: `kof-{clusterName}`
- Child ConfigMap: `kof-cluster-config-{clusterName}`
- vmrules ConfigMap: `kof-record-vmrules-{clusterName}`
- VMUser Secret: `storage-vm-user-credentials-{hashedName}`
- Propagation MCS: `kof-config-{clusterName}` / `kof-vmrules-{clusterName}`

Name hashing in `kof-operator/internal/names/` uses FNV/Adler-32. The Helm `adler32sum` helper in charts must produce matching names — **keep them in sync**.

---

## Istio Mode

Replaces Envoy Gateway, cert-manager, and external-dns with Istio service mesh. Enabled by labelling the `kof` namespace with `istio-injection: enabled`.

**Cluster labels**: `k0rdent.mirantis.com/istio-role: member` (all Istio clusters); `k0rdent.mirantis.com/istio-gateway: "true"` (regional east-west gateways).

### Parallel MCS Tracks

> Both `kof-regional` and `kof-child` render two MCS objects — standard and Istio — with non-overlapping selectors.

Reviewer: ```suggestion
### Mutually Exclusive MCS Tracks

Both `kof-regional` and `kof-child` render either standard or Istio MCS object with non-overlapping selectors.
```

| | Standard | Istio |
|---|---|---|
| Regional MCS | `kof-regional-cluster` | `kof-istio-regional-cluster` |
| Child MCS | `kof-child-cluster` | `kof-istio-child-cluster` |

> 

Reviewer: ```suggestion

In M2M (Management to Management) and M2R (Management to Regional) modes additional `kof-mgmt-cluster` or `kof-istio-mgmt-cluster` MCS is created by `kof-child` chart.
```

| Add a new storage endpoint | Controller endpoint maps, regional ConfigMap fields, both standard and Istio MCS templates |
| Add a new env var | `kof-operator/internal/env/`, `charts/kof-mothership/values.yaml`, operator Deployment template |
| Add a new controller | `kof-operator/cmd/` registration, RBAC in `charts/kof-mothership/templates/kof-operator/` |
| Add a new chart | `charts/kof/templates/` (HelmRelease + HelmChart), `charts/kof/values.yaml` (defaults + dependsOn) |
| Change VMUser credential structure | `kof-operator/internal/controller/vmuser/`, `kof-storage` VMAuth/VMUser templates, `kof-collectors` values, `kof-child` MCS `templateResourceRefs` |
| Change storage endpoint URL scheme | Controller endpoint derivation, `kof-operator/internal/env/`, `kof-storage` VMAuth route config, Istio endpoint map |
| Change recording rules format | AlertsConfigMapReconciler, `kof-storage` `valuesFrom`, `kof-regional` MCS `templateResourceRefs` |
| Change cluster name hashing | `kof-operator/internal/names/`, Helm `adler32sum` helpers (must stay in sync) |

---

## Testing and Validation

```bash
# Unit tests

> cd kof-operator && go test ./...

Reviewer: To also run `vet`, etc:

```suggestion
cd kof-operator && make test
```

---

## Key Locations Quick Reference

| Location | Purpose |
|---|---|
| `kof-operator/cmd/` | Controller registration, manager startup |
| `kof-operator/internal/env/` | Operator env var definitions |
| `kof-operator/internal/models/labels/` | Label/annotation key constants |
| `kof-operator/internal/names/` | Stable name hashing (must match Helm `adler32sum`) |
| `kof-operator/api/v1beta1/` | KOF CRD type definitions |
| `kof-operator/internal/controller/` | All reconciler implementations, endpoint maps |
| `charts/kof/values.yaml` | Top-level umbrella config (single source of truth) |
| `charts/kof/templates/` | FluxCD HelmRelease/HelmChart objects, install order |
| `charts/kof-mothership/values.yaml` | Mothership component configuration |
| `charts/kof-mothership/templates/kof-operator/` | Operator RBAC and Deployment |
| `charts/kof-mothership/templates/istio/` | Istio namespace bootstrapping |
| `charts/kof-regional/templates/` | Standard and Istio MCS for regional clusters |
| `charts/kof-child/templates/` | Standard and Istio MCS for child clusters |
| `charts/kof-storage/templates/istio/` | Istio in-mesh VMAuth Service |

> | `charts/kof-collectors/values.yaml` | Collector configuration |

Reviewer: Please ask AI to merge this with "Repository Layout" section to avoid duplication.
